snap-docs: document asynchronous changes to snapctl (#260)
canonical/snap-docs: docs/how-to-guides/snap-development/use-snapctl.md

If these commands are run from a {ref}`hook <reference-development-supported-snap-hooks>`, the components will be installed/removed after the hook itself has run if it ended successfully.

> ## snapctl commands

Author: First bit almost repeats section above, omit for clarity

> ### Asynchronous Operation

Author: I couldn't see anywhere else that obviously describes snapctl operation so I feel this may be the most appropriate place in the docs for this

If these commands are run from a {ref}`hook <reference-development-supported-snap-hooks>`, the components will be installed/removed after the hook itself has run if it ended successfully.

### Asynchronous Operation

> By default installation and removal commands are run synchronously and will block until the change is completed.

Reviewer: ```suggestion
By default, the `install` and `remove` subcommands run synchronously and will block until the change they create is completed.
```

To enable deeper insight of the change, the `--no-wait` flag can be passed. In this case, the change-id is returned immediately and the user can poll `snapctl is-ready` to determine when the change has finished.

> To view all tasks associated with a change-id in the same snapctl context, the command `snapctl tasks` may be used. Optionally, the `--format` flag may be passed with the `json` argument to return the data as JSON rather than a tabular representation.

Reviewer: What about this:

The `--no-wait` flag can be used with `install` and `remove` to indicate that `snapctl` should not block on the created change. Instead, `snapctl` will immediately return the change ID. This has similar semantics to `--no-wait` when used with `snap install/refresh`. This change ID should be used with the `is-ready` and `tasks` subcommands.

The `is-ready` subcommand enables polling the given change until it has reached a terminal state. See `snapctl is-ready --help` for a detailed breakdown of the exit codes used.  

The `tasks` subcommand prints detailed information about the given change. This includes information such as the exact state of the change and the progress of individual tasks within the change. Provide the `--format=json` argument to receive JSON output, rather than the default tabular representation.

Note that both `snapctl is-ready` and `snapctl tasks` can only be used with changes that were created by the snap in which they are being invoked. Arbitrary changes on the system cannot be introspected via these subcommands.

Author: This layout reads a lot better, and it makes sense to link --no-wait to the snap install and remove commands, ill change it.

> If these commands are run from a {ref}`hook <reference-development-supported-snap-hooks>`, the components will be installed/removed after the hook itself has run if it ended successfully.

Reviewer 2: What happens with `--no-wait` if used within a hook? Or rather, what happens if one does `snapctl is-ready <chg-id>` for an install/remove change which doesn't start until after the hook ends?

Author: I think the hook runs at the end of the change, so in one of the spread tests we make the hook wait for a file creation, the change is visible in this window.

What this is trying to say is that if its waiting on a hook to complete for install/remove, is-ready will return in progress.

Does this make sense

Reviewer 2: I think it makes sense that `is-ready` would always return in progress, because as I understand it the `snapctl install` task wouldn't actually run until after the hook finishes. I'm not sure if the test you're describing exercises this though, since the hook doesn't do any `snapctl install/remove` iirc.

### Asynchronous Operation

> By default, the `install` and `remove` sub-commands run synchronously and will block until the change they create is completed.

Reviewer 2: ```suggestion
By default, the `install` and `remove` sub-commands run synchronously and will block until the change they create is completed (unless run within a hook).
```

Reviewer 2: This is the more interesting bit though, I'm very curious what happens here. If you `snapctl install` within a hook, does it return immediately and then do the actual install after the hook finishes? How does `--no-wait` affect that behavior (if at all) when it's within a hook?

Reviewer: `--no-wait` isn't allowed when run from a hook. We should document that here.

https://github.com/canonical/snapd/blob/e1bc623cd030955f6c0986167b7da9de6394c8cc/overlord/hookstate/ctlcmd/helpers.go#L621

Reviewer: > If you snapctl install within a hook, does it return immediately and then do the actual install after the hook finishes?

The tasks are queued at the end of the change that is running the hooks.

Author: Ive documented this in the docs and source code.

> The --no-wait flag can be used with install and remove to indicate that snapctl should not block on the created change. Instead, snapctl will immediately return the change ID. This has similar semantics to --no-wait when used with snap install/refresh. This change ID should be used with the is-ready and tasks sub-commands.

Reviewer: Make sure to put arguments in backticks. See how `--default` and `--with` are used above.

> The `--no-wait` flag can be used with `install` and `remove` to indicate that snapctl should not block on the created change. Instead, snapctl will immediately return the change ID. This has similar semantics to `--no-wait` when used with `snap install/refresh`. This change ID should be used with the `is-ready` and `tasks` sub-commands. Note that `--no-wait` is not allowed to be run within a hook.

Reviewer 2: A bit of a nitpick, though maybe this is too much information:

```suggestion
The `--no-wait` flag can be used with `install` and `remove` to indicate that snapctl should not block on the created change. Instead, snapctl will immediately return the change ID. This has similar semantics to `--no-wait` when used with `snap install/refresh`. This change ID should be used with the `is-ready` and `tasks` sub-commands. Note that `--no-wait` is not allowed to be used within a hook, as `install`/`remove` from within a hook already attaches the install/remove task to the existing hook change and returns immediately.
```

Author: I think used is a better substitute for run, but I think the bit about attaching the task to the existing hook change may be too much.

Reviewer 2: Sounds reasonable to me!
